helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java

```diff
@@ -263,7 +251,11 @@ public void updatePreviousAssignedTasksStatus(
         case TASK_ABORTED:
 
         case ERROR: {
-          donePartitions.add(pId); // The task may be rescheduled on a different instance.
+          // First make this task which is in terminal state to be dropped.
+          // Later on, in next pipeline in handleAdditionalAssignments, the task will be retried if possible.
+          // (meaning it is not ABORTED and max number of attempts has not been reached yet)
+          assignedPartitions.get(instance).add(pId);
+          paMap.put(pId, new PartitionAssignment(instance, TaskPartitionState.DROPPED.name()));
```

Comment:
Again, I am not sure if this would cause the JobContext to have the wrong state for the task partitions. Did you check? What state does the task go into between being dropped and getting rescheduled? (It should be ERROR or TASK_ABORTED, or whatever it was before being dropped)

Reply:
The answer to this question is also similar to COMPLETED case :). Please see it above.